File: backend/src/main/java/com/codeloom/backend/dao/testcase/TestCaseRepositoryCustomImpl.java
package com.codeloom.backend.dao.testcase;

import static com.codeloom.backend.jooq.Tables.TEST_CASES;

import com.codeloom.backend.model.TestCase;
import java.util.ArrayList;
import java.util.Collection;
import java.util.List;
import lombok.RequiredArgsConstructor;
import org.jooq.Condition;
import org.jooq.DSLContext;
import org.springframework.stereotype.Repository;

@Repository
@RequiredArgsConstructor
public class TestCaseRepositoryCustomImpl implements TestCaseRepositoryCustom {
    private final DSLContext dsl;

    private List<Condition> conditions(long problemId, Boolean publishedOnly) {
        List<Condition> conditions = new ArrayList<>();
        conditions.add(TEST_CASES.PROBLEM_ID.eq((int) problemId));

        if (publishedOnly != null) {
            conditions.add(TEST_CASES.IS_PUBLIC.eq(publishedOnly));
        }

        return conditions;
    }

    public Collection<TestCase> findAllByProblemId(long problemId, Boolean publishedOnly) {
        return dsl.select(TEST_CASES.asterisk())
                .from(TEST_CASES)
                .where(conditions(problemId, publishedOnly))
                .fetchInto(TestCase.class);
    }

    public int countAllByProblemId(long problemId, Boolean publishedOnly) {
        Integer n = dsl.selectCount()
                .from(TEST_CASES)
                .where(conditions(problemId, publishedOnly))
                .fetchOne(0, Integer.class);
        return n == null ? 0 : n;
    }
}

File: backend/src/main/kotlin/com/codeloom/backend/dao/testcase/TestCaseRepositoryCustomImpl.kt
package com.codeloom.backend.dao.testcase

import com.codeloom.backend.jooq.tables.references.TEST_CASES
import com.codeloom.backend.model.TestCase
import org.jooq.DSLContext
import org.springframework.stereotype.Repository

@Repository
class TestCaseRepositoryCustomImpl(
    private val dsl: DSLContext,
) : TestCaseRepositoryCustom {
    override fun findAllByProblemId(
        problemId: Long,
        isPublic: Boolean?,
    ): Collection<TestCase> {
        val conditions =
            mutableListOf(
                TEST_CASES.PROBLEM_ID.eq(problemId.toInt()),
            )
        isPublic?.let {
            conditions.add(TEST_CASES.IS_PUBLIC.eq(isPublic))
        }
        val stmt =
            dsl.select(TEST_CASES.asterisk())
                .from(TEST_CASES)
                .where(conditions)
        return stmt.fetchInto(TestCase::class.java)
    }

    override fun countAllByProblemId(problemId: Long, isPublic: Boolean?): Int {
        val conditions =
            mutableListOf(
                TEST_CASES.PROBLEM_ID.eq(problemId.toInt()),
            )
        isPublic?.let {
            conditions.add(TEST_CASES.IS_PUBLIC.eq(isPublic))
        }
        val stmt =
            dsl.selectCount()
                .from(TEST_CASES)
                .where(conditions)
        return stmt.fetchOne(0, Int::class.java) ?: 0
    }
}
